homeassistant/components/elmo_alarm/__init__.py

```diff
@@ -0,0 +1,237 @@
+        self.zones = None
+        self.inputs = None
+
+        ElmoClient.__init__(self, host, vendor)
+
+    async def update(self):
+        """Get updates and refresh internal states."""
+        try:
+            data = self.check()
+        except PermissionDenied as exception:
+            _LOGGER.warning("Invalid session, trying to authenticate %s", exception)
+            try:
+                self.auth(self._username, self._password)
+                data = self.check()
+            except PermissionDenied as exception:
+                _LOGGER.warning("Invalid credentials: %s", exception)
+            except HTTPError as exception:
+                _LOGGER.warning(
+                    "Got HTTP error when authenticating. Check credentials. Code: %s",
+                    exception,
+                )
+
+        if self._data is None:
+            await self._configure_states()
+        self._data = data
+        await self._update_arm_state()
+        await self._update_zone_state()
+        await self._update_input_state()
+
+    async def _configure_states(self):
+        """Initialize the elmo alarm states."""
+        state_config = {}
+        for state in self._states_config:
+            state_config[state[CONF_NAME]] = state[CONF_ZONES]
+        self.states = state_config
+
+    async def _update_arm_state(self):
+        """Update the elmo alarm states."""
+        areas_armed = self._data["areas_armed"]
+
+        if not areas_armed:
+            self.state = STATE_ALARM_DISARMED
+        else:
+            armed_indexes = [area_data["index"] + 1 for area_data in areas_armed]
+
+            state = [
+                state
+                for state, areas in self.states.items()
+                if set(areas) == set(armed_indexes)
+            ]
+
+            if state:
+                self.state = state[0]
+            else:
+                self.state = None
+
+    async def _update_zone_state(self):
+        """Update the elmo alarm zone's states."""
+        self.zones = [
+            ZoneData(zone_id=area["index"], zone_name=area["name"], state=ZONE_ARMED)
+            for area in self._data["areas_armed"]
+            if area["name"] != "Unknown"
+        ]
+        self.zones += [
+            ZoneData(zone_id=area["index"], zone_name=area["name"], state=ZONE_DISARMED)
+            for area in self._data["areas_disarmed"]
+            if area["name"] != "Unknown"
+        ]
+
+    async def _update_input_state(self):
+        """Update the elmo alarm input's states."""
+        self.inputs = [
+            InputData(input_id=inp["index"], input_name=inp["name"], state=INPUT_ALERT)
+            for inp in self._data["inputs_alerted"]
+            if inp["name"] != "Unknown"
+        ]
+        self.inputs += [
+            InputData(input_id=inp["index"], input_name=inp["name"], state=INPUT_WAIT)
+            for inp in self._data["inputs_wait"]
+            if inp["name"] != "Unknown"
+        ]
```

Comment:
What do you think about using the filter function?

```python
    self.inputs = []

    inputs_alert = list(filter(lambda inp:(inp["name"] != "Unknown"), self._data["inputs_alerted"]))
    inputs_wait = list(filter(lambda inp:(inp["name"] != "Unknown"), self._data["inputs_wait"]))

    for input in inputs_alert:
        self.inputs.append(InputData(input_id=input["index"], input_name=input["name"], state=INPUT_ALERT))

    for input in inputs_wait:
        self.inputs.append(InputData(input_id=input["index"], input_name=input["name"], state=INPUT_WAIT))
```

Reply:
Yes, I like it. Used the same approach in _update_zone_state method.